source/common/tcp/conn_pool.h

```diff
@@ -0,0 +1,203 @@
+    }
+  }
+
+  void closeConnections() override {
+    for (auto* list : {&ready_clients_, &busy_clients_, &connecting_clients_}) {
+      while (!list->empty()) {
+        list->front()->close();
+      }
+    }
+  }
+  ConnectionPool::Cancellable* newConnection(Tcp::ConnectionPool::Callbacks& callbacks) override {
+    return Envoy::ConnectionPool::ConnPoolImplBase::newStream(reinterpret_cast<void*>(&callbacks));
+  }
+
+  ConnectionPool::Cancellable* newPendingRequest(void* context) override {
+    auto* callbacks = reinterpret_cast<Tcp::ConnectionPool::Callbacks*>(context);
+    Envoy::ConnectionPool::PendingRequestPtr pending_request(
+        new TcpPendingRequest(*this, *callbacks));
+    pending_request->moveIntoList(std::move(pending_request), pending_requests_);
+    return pending_requests_.front().get();
+  }
+
+  Upstream::HostDescriptionConstSharedPtr host() const override { return host_; }
+
+  Envoy::ConnectionPool::ActiveClientPtr instantiateActiveClient() override {
+    return Envoy::ConnectionPool::ActiveClientPtr{
+        new ActiveTcpClient(*this, host_->cluster().maxRequestsPerConnection(), 1)};
+  }
+
+  void attachRequestToClient(Envoy::ConnectionPool::ActiveClient& client,
+                             Envoy::ConnectionPool::PendingRequest& request) override {
+    TcpPendingRequest* tcp_request = reinterpret_cast<TcpPendingRequest*>(&request);
+    attachRequestToClientImpl(client, reinterpret_cast<void*>(&tcp_request->callbacks_));
+  }
+
+  void onPoolReady(Envoy::ConnectionPool::ActiveClient& client, void* context) override {
+    ActiveTcpClient* tcp_client = reinterpret_cast<ActiveTcpClient*>(&client);
+    auto* callbacks = reinterpret_cast<Tcp::ConnectionPool::Callbacks*>(context);
+    std::unique_ptr<Envoy::Tcp::ConnectionPool::ConnectionData> connection_data =
+        Envoy::Tcp::ConnectionPool::ConnectionDataPtr{
+            new ActiveTcpClient::TcpConnectionData(*tcp_client, *tcp_client->connection_)};
+    callbacks->onPoolReady(std::move(connection_data), tcp_client->real_host_description_);
+  }
+
+  void onPoolFailure(const Upstream::HostDescriptionConstSharedPtr& host_description,
+                     absl::string_view, ConnectionPool::PoolFailureReason reason,
+                     void* context) override {
+    auto* callbacks = reinterpret_cast<Tcp::ConnectionPool::Callbacks*>(context);
+    callbacks->onPoolFailure(reason, host_description);
+  }
+
+  // These two functions exist for testing parity between old and new Tcp Connection Pools.
+  virtual void onConnReleased(Envoy::ConnectionPool::ActiveClient& client) {
+    if (client.state_ == Envoy::ConnectionPool::ActiveClient::State::BUSY) {
+      if (!pending_requests_.empty() && !upstream_ready_enabled_) {
+        upstream_ready_cb_->scheduleCallbackCurrentIteration();
+      }
+    }
+  }
+  virtual void onConnDestroyed() {}
+
+protected:
+  Event::SchedulableCallbackPtr upstream_ready_cb_;
+  bool upstream_ready_enabled_ = false;
```

Comment:
I think we've been using `{false}` more commonly, instead of `= false` for in-class init.